### DISCUSSION TITLE
Numdux — a local-first data cleaning & ML notebook (feedback welcome)

### DISCUSSION THREAD

**Discussion by Kahabk** [community/community#203130] [Discussions]
### Discussion Type

Product Feedback

### Discussion Content

Hey everyone 👋

I've spent the last while building Numdux Notebook — a local-first tool for cleaning up messy datasets and training models, without shipping your data off to someone else's server.

The basic idea: you upload a dataset, it tells you what's actually wrong with it (missing values, duplicates, weird outliers, correlations, a 0–100 quality score), and from there you walk through cleaning it, engineering features, training a model, and exporting a report — all running locally, sandboxed on your own machine.

To be clear on where it's at: it's not "one click and done" yet. You go through it step by step — you see what's happening at each stage and approve what gets kept. That's actually deliberate for now, since I'd rather you trust what it did than have it silently do things to your data.

[Repo:](https://github.com/Kahabk/numdux)
<img width="1920" height="1080" alt="NUMDUX" src="https://github.com/user-attachments/assets/d196a187-f3f9-41d6-a331-50578315bf16" />

Where it's at right now:

Reads CSV, Excel, Parquet, JSON, JSONL
Auto-generates a data quality profile with charts
A guided 10-stage pipeline: Load, Explore, Visualize, Clean, Feature Engineering, Prepare, Split, Train, Tune & Evaluate, Save & Predict
Python and SQL notebook cells, run inside a Docker sandbox
Cleaning plans from a free rule engine, or Gemini if you want the AI version
A Model Lab for training and comparing scikit-learn models
Everything versioned, exportable to PDF, CSV, or Parquet

Where I want to take it: a fully automatic "Auto-Pilot" mode — upload and it does the whole pipeline on its own, no clicking through each stage — with a review step at the end so you still approve before anything's saved. Further out, a market-simulation feature to get a read on how a product might land before actually launching it.

It's fully open source, and I'd genuinely love for people to poke at it. If you:

Try it and hit something broken, tell me
Know React/Vite, FastAPI/pandas/sklearn, or sandboxing/security and want to contribute — I'd really welcome the help
Think the idea is missing something obvious — say that too

**Comment by github-actions** [comment]
**💬 Your Product Feedback Has Been Submitted 🎉**

Thank you for taking the time to share your insights with us! Your feedback is invaluable as we build a better GitHub experience for all our users.

**Here's what you can expect moving forward ⏩**
- Your input will be carefully reviewed and cataloged by members of our product teams. 
  - Due to the high volume of submissions, we may not always be able to provide individual responses.
  - Rest assured, your feedback will help chart our course for product improvements.
- Other users may engage with your post, sharing their own perspectives or experiences. 
- GitHub staff may reach out for further clarification or insight. 
  - We may 'Answer' your discussion if there is a current solution, workaround, or roadmap/changelog post related to the feedback.

**Where to look to see what's shipping 👀**
- Read the [Changelog](https://github.blog/changelog/) for real-time updates on the latest GitHub features, enhancements, and calls for feedback.
- Explore our [Product Roadmap](https://github.com/orgs/github/projects/4247), which details upcoming major releases and initiatives.

**What you can do in the meantime 💻**
- Upvote and comment on other user feedback Discussions that resonate with you.
- Add more information at any point! Useful details include: use cases, relevant labels, desired outcomes, and any accompanying screenshots.

As a member of the GitHub community, your participation is essential. While we can't promise that every suggestion will be implemented, we want to emphasize that your feedback is instrumental in guiding our decisions and priorities.

Thank you once again for your contribution to making GitHub even better! We're grateful for your ongoing support and collaboration in shaping the future of our platform. ⭐

**Comment by ebndev** [comment]
Hi @Kahabk, thanks for sharing with the community! 

Unfortunately, we currently do not allow self-promotion, advertising, or solicitation in Community Discussions. We want to make sure there is space for users to ask questions without overwhelming them with other conversations. Thank you for helping us maintain a productive and tidy community for all our members.
